package org.grakovne.lissen.ui.screens.settings.composable

import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Check
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.ListItem
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import org.grakovne.lissen.ui.components.LissenModalBottomSheet
import org.grakovne.lissen.ui.theme.tvFocusIndication

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CommonSettingsItemComposable(
  items: List<CommonSettingsItem>,
  selectedItem: CommonSettingsItem?,
  onDismissRequest: () -> Unit,
  onItemSelected: (CommonSettingsItem) -> Unit,
  selectedImage: ImageVector = Icons.Outlined.Check,
) {
  var activeItem by remember { mutableStateOf(selectedItem) }

  LissenModalBottomSheet(
    containerColor = MaterialTheme.colorScheme.background,
    scrollable = false,
    onDismissRequest = onDismissRequest,
    content = {
      Column(
        modifier =
          Modifier
            .fillMaxWidth()
            .padding(bottom = 16.dp)
            .padding(horizontal = 16.dp),
      ) {
        Spacer(modifier = Modifier.height(8.dp))

        LazyColumn(modifier = Modifier.fillMaxWidth()) {
          itemsIndexed(items) { index, item ->
            ListItem(
              leadingContent = {
                item.icon?.let {
                  Icon(
                    imageVector = it,
                    contentDescription = null,
                    modifier = Modifier.size(24.dp),
                  )
                }
              },
              headlineContent = {
                Row {
                  Text(
                    text = item.name,
                    color =
                      when (item.enabled) {
                        true -> MaterialTheme.colorScheme.onSurface
                        false -> MaterialTheme.colorScheme.onSurface.copy(alpha = 0.4f)
                      },
                  )
                }
              },
              supportingContent =
                item.description?.let {
                  {
                    Text(
                      text = it,
                      style = MaterialTheme.typography.bodyMedium,
                      color =
                        when (item.enabled) {
                          true -> MaterialTheme.colorScheme.onSurfaceVariant
                          false -> MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.4f)
                        },
                    )
                  }
                },
              trailingContent = {
                if (item.id == activeItem?.id) {
                  Icon(
                    imageVector = selectedImage,
                    contentDescription = null,
                    modifier = Modifier.size(24.dp),
                  )
                }
              },
              modifier =
                Modifier
                  .fillMaxWidth()
                  .clickable(
                    enabled = item.enabled,
                    indication = tvFocusIndication(),
                    interactionSource = remember { MutableInteractionSource() },
                  ) {
                    activeItem = item
                    onItemSelected(item)
                  },
            )
            if (index < items.lastIndex) {
              HorizontalDivider()
            }
          }
        }
      }
    },
  )
}
